File: base/src/main/java/io/spine/base/EntityState.java
/*
 * Copyright 2021, TeamDev. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Redistribution and use in source and/or binary forms, with or without
 * modification, must retain the above copyright notice and the following
 * disclaimer.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
 * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
 * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
 * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
 * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
 * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
 * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

package io.spine.base;

import com.google.errorprone.annotations.Immutable;
import io.spine.annotation.GeneratedMixin;
import io.spine.validate.BuiltMessage;
import io.spine.validate.ValidatingBuilder;

/**
 * A common interface for entity state messages.
 *
 * <p>Any message that defines an {@code (entity)} option with a valid {@code kind} is marked with
 * this interface by the Model Compiler.
 *
 * <p>The first field of the entity state message is treated as its identifier. It is a convention
 * that has two goals:
 *
 * <ol>
 *     <li>The definition of an entity state always starts with its ID with no extra Protobuf
 *     options. This way it's feels easy and more natural to read the code.

 *     <li>Developers don't forget to specify which of the fields declared in Protobuf corresponds
 *     to the entity ID.
 * </ol>
 *
 * <p>At codegen-time, the Model Compiler substitutes the generic parameter {@code <I>} with
 * an actual type of the first field of the entity state message.
 *
 * @param <I>
 *         the type of entity identifiers
 * @param <B>
 *         the type of the builder for this entity state
 * @param <M>
 *         the type of this entity state, which is used for binding of the builder type
 * @see io.spine.code.proto.EntityStateOption
 */
@SuppressWarnings("unused") /* Used in the generated code. */
@Immutable
@GeneratedMixin
public interface EntityState<I, B extends ValidatingBuilder<M>, M extends EntityState<I, B, M>>
        extends BuiltMessage<B, M> {

}

File: base/src/main/kotlin/io/spine/base/EntityState.kt
/*
 * Copyright 2025, TeamDev. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * https://www.apache.org/licenses/LICENSE-2.0
 *
 * Redistribution and use in source and/or binary forms, with or without
 * modification, must retain the above copyright notice and the following
 * disclaimer.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
 * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
 * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
 * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
 * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
 * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
 * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

package io.spine.base

import com.google.errorprone.annotations.Immutable

/**
 * A common interface for entity state messages.
 *
 * Any message that defines an `(entity)` option with a valid `kind` is an entity state.
 *
 * The `kind` of the entity state defines the actual interface of the entity state
 * which could be this interface or one extends it.
 * If the `kind` is [ENTITY][io.spine.option.EntityOption.Kind.ENTITY] the generated
 * message will implement this interface. Otherwise, it would be one of the subinterfaces.
 * For example, if the `kind` is [PROJECTION][io.spine.option.EntityOption.Kind.PROJECTION],
 * the generated message will implement [ProjectionState].
 *
 * The field of the entity state message, which is declared first in the code of
 * the message, is treated as its identifier. The "first" field is determined by the order
 * in which fields appear in the Protobuf message definition (reading order from top to bottom),
 * not by the field number. For example:
 *
 * ```
 * message UserView {
 *     option (entity).kind = PROJECTION;
 *
 *     UserName name = 2;  // This is the ID field (first in reading order)
 *     UserId id = 1;       // Not the ID field (second in reading order)
 * }
 * ```
 *
 * This convention has two goals:
 *
 *  1. The definition of an entity state always starts with its ID with no
 *   extra Protobuf options.
 *
 *  2. Developers don't forget to specify which of the fields declared in Protobuf
 *   corresponds to the entity ID.
 *
 * The reading order approach also provides benefits:
 *
 *  - Easier to read and understand — developers see the ID field immediately.
 *  - Supports field deprecation scenarios — if an ID field needs to be replaced
 *    (e.g., upgrading from `int32` to `int64`), the new field can be added at the top
 *    while the old field is deprecated in place.
 *
 * During code generation, the CoreJvm Compiler substitutes the generic parameter
 * [I] with an actual type of the first field of the entity state message.
 *
 * @param I The type of entity identifiers.
 *   It could be one of the types described in the [IdType] enum.
 */
@Immutable
@Suppress("unused" /* The parameter type <I> is meant to be used in the generated code. */)
public interface EntityState<I : Any> : Routable

/**
 * An entity state of an aggregate.
 *
 * Messages that have `(entity)` option with `kind` set to
 * [AGGREGATE][io.spine.option.EntityOption.Kind.AGGREGATE]
 * are instances of this class.
 */
public interface AggregateState<I : Any> : EntityState<I>

/**
 * An entity state of a projection.
 *
 * Messages that have `(entity)` option with `kind` set to
 * [PROJECTION][io.spine.option.EntityOption.Kind.PROJECTION]
 * are instances of this interface.
 *
 * @param I The type of the projection identifiers.
 */
public interface ProjectionState<I : Any> : EntityState<I>

/**
 * Same as [ProjectionState].
 */
public typealias ViewState<I> = ProjectionState<I>

/**
 * An entity state of a process manager.
 *
 * Messages that have `(entity)` option with `kind` set to
 * [PROCESS_MANAGER][io.spine.option.EntityOption.Kind.PROCESS_MANAGER]
 * are instances of this interface.
 */
public interface ProcessManagerState<I : Any> : EntityState<I>
